galley/pkg/runtime/resource/resource.go

```diff
@@ -0,0 +1,107 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package resource contains core abstract types for representing configuration resources.
+package resource
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/gogo/protobuf/proto"
+)
+
+// Kind is the name of a resource type.
+type Kind string
```

Comment:
Is this k8s kind or protobuf type_url?

Reply:
Looks like this is the protobuf MessageName based on `NewProtoInstance` implementation. It would be useful for us to use consistent names across packages (when possible) to differentiate the various forms of configuration, e.g.

- **kind**: k8s kind, e.g. VirtualService
- **messageName**: name of the protobuf message for use with protobuf.MessageType() and other protobuf functions, e.g. istio.io.networking.v1alpha3.VirtualService). 
- **typeURL**: the fully qualified resource name using in `any` messages over the wire and with MCP.

Reply:
I've gone back and forth on this a bit. As the code stands, resource.Kind is the messageName, but it should be something more that is independent of both the source and destination types, possibly including versioning aspects as well. I can rename it to something else. I don't mind standardizing on the names you suggested (most of the code should be on that plan anyways, except resource.Kind).

In any case, it should not escape the protocol boundaries and stay in Galley. We have typeURLs and CRD kinds on either side.